buildstockbatch/postprocessing.py

```diff
@@ -292,6 +303,81 @@ def correct_schema(cur_schema_dict, df):
+    """
+        Returns intelligent grouping of building_ids by partition columns.
+        1. Group the building_ids by partition columns. For each group, say (CO, Jefferson), we have a list of building
+           ids. The total number of such groups is ngroups
+        2. Concatenate those list to get bldg_id_list, which will have all the bldg_ids but ordered such that that
+           buildings belonging to the same group are close together.
+        3. Split the list of building in each group in 1 to multiple subgroups so that total number of buildings
+           in each group is less than or equal to files_per_partition. This will give the bldg_id_groups (list of list)
+           used to read the dataframe. The buildings within the inner list will be concatenated. len(bldg_id_groups) is
+           equal to number of such concatenation, and eventually, number of output parquet files.
+    """
+    total_building = len(partition_map)
+    bldg_id_list_df = partition_map.reset_index().groupby(partition_columns)['building_id'].apply(list)
+    ngroups = len(bldg_id_list_df)
+    bldg_id_list = bldg_id_list_df.sum()
+    nfiles_in_each_group = [nfiles for nfiles in bldg_id_list_df.map(lambda x: len(x))]
+    files_groups = [split_into_groups(n, files_per_partition) for n in nfiles_in_each_group]
+    flat_groups = [n for group in files_groups for n in group]
+    cum_files_count = np.cumsum(flat_groups)
+    assert cum_files_count[-1] == total_building
+    cur_index = 0
+    bldg_id_groups = []
+    for indx in cum_files_count:
+        bldg_id_groups.append(bldg_id_list[cur_index:indx])
+        cur_index = indx
+
+    return bldg_id_groups, bldg_id_list, ngroups
+
+
+def get_bldg_groups(bldg_id_list, files_per_partition):
+    """
+        Similar to get_partitioned_buildings_groups, but without the partition columns. Groups buildings into sub list
+        of max of files_per_partition per group.
+    """
+    bldg_id_list = list(bldg_id_list)
+    total_building = len(bldg_id_list)
+    nfiles_in_each_group = split_into_groups(total_building, files_per_partition)
+    cum_files_count = np.cumsum(nfiles_in_each_group)
+    assert cum_files_count[-1] == total_building
+    bldg_id_groups = []
+    cur_index = 0
+    for indx in cum_files_count:
+        bldg_id_groups.append(bldg_id_list[cur_index:indx])
+        cur_index = indx
+
+    return bldg_id_groups, bldg_id_list, 1
+
+
+def get_upgrade_list(cfg):
+    upgrade_start = 1 if cfg['baseline'].get('skip_sims', False) else 0
+    upgrade_end = len(cfg.get('upgrades', [])) + 1
+    full_upgrade_list = range(upgrade_start, upgrade_end)
+    upgrade_list = cfg.get('postprocessing', {}).get('upgrades', [])
+    upgrade_list = sorted(upgrade_list) or full_upgrade_list
+    for upgrade in upgrade_list:
+        if upgrade not in full_upgrade_list:
+            logger.warning(f"Upgrade {upgrade} specified in the yaml is not valid")
```

Comment:
I'm a little confused here. Wouldn't we want it to error out if an upgrade is missing rather than just rolling on by?